### ISSUE TITLE
Par2 fails to validate file where block size is large.

### ISSUE THREAD

**Issue by exeter-matthew-wakeling** [Parchive/par2cmdline#118]
I have a 357GB file that I wish to archive, so I ran the following command:

par2 c -b100 -c2 -n1 -m1000 file.par2 file

and I then have the following:

```
# ls -l
-rw-r--r-- 1 user group 382969887758 Apr 27  2017 file
-rw-r--r-- 1 user group   7698086468 Apr 27  2017 file.vol0+2.par2
```

If I then run "par2 v -q file.vol0+2.par2" then I get the following:

```
par2cmdline version 0.4, Copyright (C) 2003 Peter Brian Clements.

par2cmdline comes with ABSOLUTELY NO WARRANTY.

This is free software, and you are welcome to redistribute it and/or modify
it under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or (at your
option) any later version. See COPYING for details.

Loading "file.vol0+2.par2".
Could not read 7698081584 bytes from file at offset 0
```

However, if I run "cat file >/dev/null" I get no errors. I suspect that there is a problem with par2 trying to read >2GB in one block. I have similar problems with another large file.

**Comment by SethosII** [comment]
I also encounter errors during verification of large files. It is working for a small amount of TB but stops working for larger files:

```
> par2 c -u -r10 -m32768 large.tar.bz2
Block size: 3060274512                                                                                                                                       
Source file count: 1                                                                                                                                         
Source block count: 2000                                                                                                                                     
Redundancy: 10%                                                                                                                                              
Recovery block count: 200                                                                                                                                    
Recovery file count: 8                                                                                                                                       
                                                                                                                                                             
Opening: large.tar.bz2                                                                                                                               
Computing Reed Solomon matrix.                                                                                                                               
Constructing: done.                                                                                                                                          
Wrote 34359737600 bytes to disk                                                                                                                              
Wrote 34359737600 bytes to disk                                                                                                                              
Wrote 34359737600 bytes to disk                                                                                                                              
Wrote 34359737600 bytes to disk                                                                                                                              
Wrote 34359737600 bytes to disk                                                                                                                              
Wrote 34359737600 bytes to disk                                                                                                                              
Wrote 34359737600 bytes to disk                                                                                                                              
Wrote 34359737600 bytes to disk                                                                                                                              
Wrote 34359737600 bytes to disk                                                                                                                              
Wrote 34359737600 bytes to disk                                                                                                                              
Wrote 34359737600 bytes to disk                                                                                                                              
Wrote 34359737600 bytes to disk                                                                                                                              
Wrote 34359737600 bytes to disk                                                                                                                              
Wrote 34359737600 bytes to disk                                                                                                                              
Wrote 34359737600 bytes to disk
Wrote 34359737600 bytes to disk
Wrote 34359737600 bytes to disk
Wrote 27939363200 bytes to disk
Writing recovery packets
Writing verification packets
Done
> par2 verify large.tar.bz2
Loading "large.tar.bz2.par2".
Loaded 4 new packets
Loading "large.tar.bz2.vol000+25.par2".
Loaded 25 new packets including 25 recovery blocks
Loading "large.tar.bz2.vol025+25.par2".
Loaded 25 new packets including 25 recovery blocks
Loading "large.tar.bz2.vol050+25.par2".
Loaded 25 new packets including 25 recovery blocks
Loading "large.tar.bz2.vol075+25.par2".
Loaded 25 new packets including 25 recovery blocks
Loading "large.tar.bz2.vol100+25.par2".
Loaded 25 new packets including 25 recovery blocks
Loading "large.tar.bz2.vol125+25.par2".
Loaded 25 new packets including 25 recovery blocks
Loading "large.tar.bz2.vol150+25.par2".
Loaded 25 new packets including 25 recovery blocks
Loading "large.tar.bz2.vol175+25.par2".
Loaded 25 new packets including 25 recovery blocks

There are 1 recoverable files and 0 other files.
The block size used was 3060274512 bytes.
There are a total of 2000 data blocks.
The total size of the data files is 6120549019350 bytes.

Verifying source files:

Could not read 6120549024 bytes from /mnt/tararchive/secondary_analysis/COPD_52fb6aae.tar.bz2 at offset 0
```

The size of the file is 5.6 TB (`6120549019350` to be exact). Is there some limit exceeded here?

This is on a Ubuntu 18.04 with par2 `0.8.0` from the standard repositories.

**Comment by mdnahas** [comment]
A 32-bit integer can count to 2^32 or 4GB.  The standard uses 64-bit integers for sizes so there shouldn't be a problem with large files, but the par2 program may be using a 32-bit integer somewhere.  I'll take a look.

**Comment by mdnahas** [comment]
I experienced the same problem.  It failed with a blocksize of 7GB.  (But worked with 3GB.)

It looks like the problem is in diskfile.cpp.  There is code to determine a maximum offset and maximum read length: 

#define MaxLength 0xffffffffUL

So, the current code is limited to blocksize < 4GB.  I don't know why it should be limited on 64-bit Linux.  And, even if the read length is limited on 32-bit Linux, why the code cannot do multiple calls to fread.

I'll look into fixing it.  It may be a complicated fix, because it relates to the OS being used.

**Comment by mdnahas** [comment]
I think there's two parts to this problem.

First, on 64-bit Linux, we should support read lengths of more than 4GB.  We need to fix the calculate of MaxLength.

The other problem is that Windows's FileWrite command is limited to 32-bit unsigned length.  So, we need to support processing block sizes of larger than 32-bits, using reads of size under 32-bits.  This means exposing the MaxLength value of DiskFile and making sure the "chunksize" in par2creator and par2repairer is less than the chunksize.  

I'll try to fix the second problem.   It will ship after the big change is pushed out.

SIDENOTE #1: It's cool that people are trying terabytes and we're blowing out a 32-bit value! 
SIDENOTE #2: I'm thinking Tornado Codes are more of a priority, if people are using terabytes!

**Comment by Rhialto** [comment]
On Fri 17 May 2019 at 10:48:17 -0700, Michael Nahas wrote:
> The other problem is that Windows's FileWrite command is limited to 32-bit unsigned length.  So, we need to support processing block sizes of larger than 32-bits, using reads of size under 32-bits.  This means exposing the MaxLength value of DiskFile and making sure the "chunksize" in par2creator and par2repairer is less than the chunksize.  

Or, and maybe this is easier, create a wrapper function for file reading/writing that has a 64-bit size. For OSes where it is needed, it will do multiple reading/writing calls with blocks that are small enough.

**Comment by mdnahas** [comment]
DiskFile has 2 implementations, one Windows and one Unix.  I was hoping to avoid touching the Windows code as much as possible, since it's harder for me to test.  But a wrapper function is definitely the easier way to go.  I'm glad I integrated MinGW+WINE. :)

**Comment by mdnahas** [comment]
I wrote the code and one test.  It's in the branch named "long_read_write" in my repo.
https://github.com/mdnahas/par2cmdline/tree/long_read_write

The change is on top of the fork with my libpar2 changes.  Once that fork is merged into the main repo, I'll push this code.

**Comment by BlackIkeEagle** [comment]
Should be fixed in master now, please re-open if issues still exist

**Comment by SethosII** [comment]
@mdnahas I tested your branch and I can confirm that this fixes my issue, thank you very much!

**Comment by exeter-matthew-wakeling** [comment]
Confirmed - I tested the new master against a file that previously failed to validate and it works now. Many thanks.

**Comment by Neustradamus** [comment]
Any news?

- https://github.com/Parchive/par2cmdline/issues/138
